src/proc/mod.rs

```diff
@@ -130,6 +132,50 @@ impl super::TypeInner {
             Self::Image { .. } | Self::Sampler { .. } => 0,
         }
     }
+
+    /// Return the canoncal form of `self`, or `None` if it's already in
+    /// canonical form.
+    ///
+    /// Certain types have multiple representations in `TypeInner`. This
+    /// function converts all forms of equivalent types to a single
+    /// representative of their class, so that simply applying `Eq` to the
+    /// result indicates whether the types are equivalent, as far as Naga IR is
+    /// concerned.
+    pub fn canonical_form(&self, types: &crate::Arena<crate::Type>) -> Option<crate::TypeInner> {
```

Comment:
Can we really have the `Eq` on types working? I know we are doing it in the validator, and it was always a little bit concerning. For example, today the IR may have 2 different types (A and B) that are `u32`, and comparing them would work because they are compared by value. But wrapping them into anything will not work, e.g `array<A, 3>` vs `array<B, 3>` would compare to false because the base type is just a handle, and the handle is different...

So this function here warrants a much larger discussion, which I filed https://github.com/gfx-rs/naga/issues/1385 for. Perhaps, at least adding a comment here saying "this only works for a specific kind of canonicalization"?

Reply:
This should be fixed by #1419.